### PR TITLE
make Lookup() return all blobs

### PR DESCRIPTION
**Pull request by aawsome**
What is the purpose of this change? What does it change?
--------------------------------------------------------

Change `repository.Index.Lookup()` to return all results.

Currently, `Lookup` returns all results including duplicates within the first index file with first match; it however does not return all duplicates. This PR "corrects" `Lookup` to return all duplicates.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

see #2523 

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have added tests for all changes in this PR (and some more)
- I have not added documentation for the changes (in the manual)
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

**Comment by greatroar** [comment]
Great! I was wanting to do this for #2788, but settled on a more local patch.

File: internal/repository/index.go
Comment (greatroar) [inline internal/repository/index.go:229]: Here again, the second return value isn't necessary. The caller can just check the len of the slice.

Comment (aawsome) [reply internal/repository/index.go]: You are right, again :smiley: 
I changed this, thanks for the suggestion!

File: internal/repository/master_index.go
Comment (greatroar) [inline internal/repository/master_index.go:45]: There's a potential for optimization here: if Index.LookupAll appends instead of always returning a new list, then this loop can be done with fewer allocations. Viz.,

```go
 func (mi *MasterIndex) LookupAll(id restic.ID, tpe restic.BlobType, blobs []restic.PackedBlob) []restic.PackedBlob {
	// lock+unlock
	for _, idx := range mi.idx {
		blobs = idx.LookupAll(id, tpe, blobs)
	}
	return blobs
}
```

File: internal/repository/master_index.go
Comment (greatroar) [inline internal/repository/master_index.go:32]: Naked return.

File: internal/repository/master_index.go
Comment (greatroar) [reply internal/repository/master_index.go:45]: See aawsome/restic#2.

Comment (aawsome) [reply internal/repository/master_index.go]: Thank you for pointing this out! However, I'm not confident to also changing `MasterIndex.LookupAll` as this would be usually not called multiple times.

I just changed `Index.LookupAll`.

File: internal/repository/master_index.go
Comment (aawsome) [reply internal/repository/master_index.go]: This is the original code ^^ - is changed now.

**Comment by greatroar** [comment]
LGTM.

**Review by MichaelEischer — COMMENTED** [review]
I've thought a bit more about the effects of adding `Lookup` and `LookupAll`. The main difference is probably that `Repository.LoadBlob` now has to traverse the full index, instead of only half on average. This is probably most noticeable for the `check` command which issues a massive amount of `LoadBlob` calls (via `LoadTree`).

Just to get an idea of how many `Lookup` calls are issued by check: I've used PR #2328 (check optimization), which already dramatically reduces the number of `LoadBlob` calls, on my large test repository which has a 7GB index with 729 files. `LoadBlob` spends roughly 40% of its runtime in the Lookup call (which amounts to more than half an hour CPU time). Using `LookupAll` would increase that to nearly 60%. That probably sounds worse than it actually is as the index lookup only amounts to 6% of the overall runtime, so that the slowdown should be tolerable.

For another repository using the standard check implementation the `Lookup`/`LookupAll` call also only requires a small part of the overall runtime.

Regarding `LookupBlobSize` I would be interested to see a small performance benchmark similar to the `BenchmarkMasterIndexLookupMultipleIndex` benchmark that already exists for the master index.

I think this PR also needs a test that `Lookup` and `LookupAll` really work as expected. The adapted tests probably don't completely cover the new semantics, especially of `LookupAll`.

**Comment by aawsome** [comment]
@MichaelEischer Maybe we should first talk about #2799 which should solve the "many index files" issue before continuing to work on this?

However, I'll take care of the points you mentioned!

File: internal/restorer/restorer.go
Comment (greatroar) [reply internal/restorer/restorer.go:316]: I don't care who does it :)

**Comment by greatroar** [comment]
Closes #2701.

File: internal/restorer/restorer.go
Comment (aawsome) [reply internal/restorer/restorer.go]: is added.

File: internal/restorer/filerestorer.go
Comment (aawsome) [reply internal/restorer/filerestorer.go]: is changed now.

**Comment by aawsome** [comment]
> Regarding `LookupBlobSize` I would be interested to see a small performance benchmark similar to the `BenchmarkMasterIndexLookupMultipleIndex` benchmark that already exists for the master index.
> 
> I think this PR also needs a test that `Lookup` and `LookupAll` really work as expected. The adapted tests probably don't completely cover the new semantics, especially of `LookupAll`.

I added this benchmark and tests. All code in `Lookup` and `LookupAll` (both in `index.go` and `master_index.go`) is now covered by the tests.

**Comment by greatroar** [comment]
LookupBlobSize isn't covered, though. The restorer tests don't hit it either (low coverage), nor do the fuse ones (they use the blobSizeCache).

As an aside, it's funny how the MasterIndex is now the only type that implements restic.Index and everything still works.

**Comment by aawsome** [comment]
> LookupBlobSize isn't covered, though. The restorer tests don't hit it either (low coverage), nor do the fuse ones (they use the blobSizeCache).

I added a bunch of easy-to-add tests to improve test coverage.
 
> As an aside, it's funny how the MasterIndex is now the only type that implements restic.Index and everything still works.

:smiley: Yes - in fact, `internal/repository/index.Index` doesn't need to implement `restic.Index`.

**Review by MichaelEischer — COMMENTED** [review]
I just took another look at this PR. Besides tests there's just a _single_ user of `Lookup` and that is inside `cmd_cat.go` which doesn't matter for performance as it's just a single call. All other uses which just need a single result blob seem to be covered by `LookupSize`.

So lets keep this simple: please drop the `Lookup` method and rename `LookupAll` to `Lookup`.

**Comment by aawsome** [comment]
> I just took another look at this PR. Besides tests there's just a _single_ user of `Lookup` and that is inside `cmd_cat.go` which doesn't matter for performance as it's just a single call. All other uses which just need a single result blob seem to be covered by `LookupSize`.
> 
> So lets keep this simple: please drop the `Lookup` method and rename `LookupAll` to `Lookup`.

You are right and especially after #2818 is merged, there shouldn't be much performance issues any more.
So this PR is now merely a code clean.

I rebased and made your suggested changes.

Seems, we have a CI issue ATM...

File: internal/repository/master_index_test.go
Comment (MichaelEischer) [inline internal/repository/master_index_test.go:64]: s/idx1/idx2/

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
I found a few further small issues. Once these are fixed, the PR should be ready for merging.

File: internal/archiver/blob_saver_test.go
Comment (aawsome) [reply internal/archiver/blob_saver_test.go:43]: `*repository.Index` does no longer implement `restic.Index`.

File: internal/repository/index.go
Comment (aawsome) [reply internal/repository/index.go:220]: Ups - that was when rebasing gave me something to resolve.. Also just realized that I kicked out the lock by accident :-1: 

Is reverted.

File: internal/repository/index.go
Comment (aawsome) [reply internal/repository/index.go:179]: is changed.

File: internal/archiver/blob_saver_test.go
Comment (aawsome) [reply internal/archiver/blob_saver_test.go:43]: Ok, just realized that `*repository.Index` does NO longer implement `restic.Index The `Lookup` method has another syntax. So I keep this change.

File: internal/repository/master_index_test.go
Comment (aawsome) [reply internal/repository/master_index_test.go]: changed.

File: internal/archiver/blob_saver_test.go
Comment (MichaelEischer) [reply internal/archiver/blob_saver_test.go:43]: I've renamed the Index interface to `MasterIndex` to make this a little less confusing.

**Review by MichaelEischer — APPROVED** [review]
LGTM. Thanks for the cleanup!

**Comment by aawsome** [comment]
@MichaelEischer Wow you are really working through the PRs today! Thanks a lot for all your work! I really appreciate a lot what you are supplying here! Also know about the burden of "the one who has to make the decission"...
:+1: 

I really like the progress restic is making. Wondering whether all the improvements done in the last month would justify a new release.. :wink:

**Comment by rawtaz** [comment]
LOL, you don't think we have thought about a new release? :P

**Comment by aawsome** [comment]
> LOL, you don't think we have thought about a new release? :P
> 
> Thing is, it's not just you guys' PRs and work that should go into a new release, but most of the developer time (primarily Michael's) is going to that instead of older PRs and stuff ;)

Sure.. is there a roadmap what points you maintainers see as important milestones or priorities towards a next release? (If there is, I just didn't find it so far). Please give me a hint if I can help with something.

**Comment by rawtaz** [comment]
Thank you @aawsome, but I was mostly thinking about having mentioned an upcoming release in a few places by now. But it's been like that for a while. We mostly have discussions amongst ourselves as part of the ongoing work to get a new release out, there's no roadmap I can point you to. It's done when it's done, and that shouldn't be too far into the future :) Thanks again!

**Pull request merged**
